File: src/main/scala/Race/evolution/Cars/SimpleCar.scala
package Race.evolution.Cars

import java.awt.Polygon

import Race.common._


object SimpleCar extends CarModel {

  def maxAccelerationDouble = 10
  def maxDeceleration: Double = 7

  def maxSteer: Double = Math.toRadians(15)

  override def drive(acceleration: Double, steer: Double, previousState: CarState): CarState = {
    val state = previousState.asInstanceOf[SimpleCarState]

    val cappedSteer = Math.max(-1d, Math.min(1d, steer))
    val newDirection =
      (cappedSteer * maxSteer) + state.direction match {
        case tooBig if tooBig > Math.PI => -Math.PI + (tooBig - Math.PI)
        case tooSmall if tooSmall < -Math.PI => Math.PI + (tooSmall + Math.PI)
        case fine => fine
      }

    val cappedAcc = Math.max(-1d, Math.min(1d, acceleration))
    val accForce = if (cappedAcc > 0) maxAccelerationDouble * cappedAcc else maxDeceleration * cappedAcc

    val (x, y) = Force.vector(accForce, newDirection)

    val newMomentum = ((state.momentum._1 / 2) + x, (state.momentum._2 / 2) + y)

    val newPosition = previousState.centerOfMass + Point(Math.round(newMomentum._1).toInt, Math.round(newMomentum._2).toInt)

    SimpleCarState(newPosition, newDirection, newMomentum)
  }

  override def initialise(track: Track): CarState = SimpleCarState(track.StartPoint, track.StartDirection, (0, 0))

  override def initialise(gridPosition: Int, track: Track): CarState = SimpleCarState(track.StartGrid(gridPosition), track.StartDirection, (0, 0))

  def rotatedCar(origin: Point, rad: Double): Polygon = {
    val points = bodyOrigins
      .map(_.rotate(rad))
      .map(_ + origin)

    new Polygon(points.map(_.x), points.map(_.y), points.length)
  }

  def lidarOrigins: List[Point] = List(Point(0, -1000), Point(-500, -1000), Point(500, -1000), Point(-1000, -700), Point(1000, -700), Point(-1000, 0), Point(1000, 0))
  def bodyOrigins: Array[Point] = Array(Point(-5, 10), Point(5, 10), Point(5, -10), Point(-5, -10))
}

File: PyCar/cars/simple_car.py
from math import radians, pi, sin, cos

from common.car import Car
from common.car_state import CarState
from common.point import Point


class SimpleCar(Car):
    def __init__(self, max_acceleration=10, max_decel=7, max_steer=radians(15), preserved_momentum=0.7):
        self.max_acceleration = max_acceleration
        self.max_decel = max_decel
        self.max_steer = max_steer
        self.preserved_momentum = preserved_momentum

    def drive(self, acceleration, steer, state: CarState) -> CarState:
        capped_steer = max(-1.0, min(1.0, steer)) * self.max_steer + state.direction

        if capped_steer > pi:
            new_direction = -pi + (capped_steer - pi)
        elif capped_steer < -pi:
            new_direction = pi + (capped_steer + pi)
        else:
            new_direction = capped_steer

        capped_acc = max(-1.0, min(1.0, acceleration))

        if capped_acc > 0:
            acc_force = self.max_acceleration * capped_acc
        else:
            acc_force = self.max_decel * capped_acc

        x, y = SimpleCar.force_vector(acc_force, new_direction)

        new_momentum = ((state.momentum[0] * self.preserved_momentum) + x,
                        (state.momentum[1] * self.preserved_momentum) + y)

        new_position = state.center_of_mass + Point(round(new_momentum[0]), round(new_momentum[1]))

        return CarState(new_position, new_direction, new_momentum)

    def lidar_origins(self):
        return [Point(0, -1000),
                Point(-500, -1000),
                Point(500, -1000),
                Point(-1000, -700),
                Point(1000, -700),
                Point(-1000, 0),
                Point(1000, 0)]

    @staticmethod
    def force_vector(magnitude: float, rad: float) -> (float, float):
        return sin(rad) * magnitude, cos(rad) * -magnitude
